-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-49702/47326

ERROR: Build errors found during clang-tidy run.
```
src/RecoLocalCalo/HcalRecProducers/plugins/alpaka/Mahi.dev.cc:659:23: error: expected ')' [clang-diagnostic-error]
  659 |                       tdcTime =
      |                       ^
src/RecoLocalCalo/HcalRecProducers/plugins/alpaka/Mahi.dev.cc:658:24: note: to match this '('
--
gmake: *** [config/SCRAM/GMake/Makefile.coderules:129: code-checks] Error 2
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

--- a/DQM/HcalCommon/interface/ValueQuantity.h
+++ b/DQM/HcalCommon/interface/ValueQuantity.h
@@ -72,6 +72,8 @@ namespace hcaldqm {
       fDiffRatio,  // (v2-v1)/v1
       fCPUenergy,
       fGPUenergy,
+      fCPUtime,
+      fGPUtime,
       fNbins,
     };
     const std::map<ValueQuantityType, std::string> name_value = {
@@ -139,6 +141,8 @@ namespace hcaldqm {
         {fDiffRatio, "(GPU energy - CPU energy)/CPU energy"},
         {fCPUenergy, "CPU energy"},
         {fGPUenergy, "GPU energy"},
+        {fCPUtime, "CPU time"},
+        {fGPUtime, "GPU time"},
         {fNbins, "Nbins"},
     };
     const std::map<ValueQuantityType, double> min_value = {
@@ -206,6 +210,8 @@ namespace hcaldqm {
         {fDiffRatio, -2.5},
         {fCPUenergy, 0},
         {fGPUenergy, 0},
+        {fCPUtime, 0},
+        {fGPUtime, 0},
         {fNbins, 0},
     };
     const std::map<ValueQuantityType, double> max_value = {
@@ -273,6 +279,8 @@ namespace hcaldqm {
         {fDiffRatio, 2.5},
         {fCPUenergy, 200},
         {fGPUenergy, 200},
+        {fCPUtime, 200},
+        {fGPUtime, 200},
         {fNbins, 50},
     };
     const std::map<ValueQuantityType, int> nbins_value = {
@@ -339,6 +347,8 @@ namespace hcaldqm {
         {fDiffRatio, 50},
         {fCPUenergy, 1000},
         {fGPUenergy, 1000},
+        {fCPUtime, 1000},
+        {fGPUtime, 1000},
         {fNbins, 50},
     };
     class ValueQuantity : public Quantity {

--- a/DQM/HcalTasks/plugins/HcalGPUComparisonTask.cc
+++ b/DQM/HcalTasks/plugins/HcalGPUComparisonTask.cc
@@ -44,6 +44,7 @@ private:
   hcaldqm::Container2D energyGPUvsCPU_subdet_;
   hcaldqm::Container1D energyDiffGPUCPU_subdet_;
   hcaldqm::ContainerProf2D energyDiffGPUCPU_depth_;
+  hcaldqm::Container2D timeGPUvsCPU_subdet_;
 };
 
 HcalGPUComparisonTask::HcalGPUComparisonTask(edm::ParameterSet const& ps)
@@ -84,10 +85,18 @@ HcalGPUComparisonTask::HcalGPUComparisonTask(edm::ParameterSet const& ps)
                                      new hcaldqm::quantity::DetectorQuantity(hcaldqm::quantity::fiphi),
                                      new hcaldqm::quantity::ValueQuantity(hcaldqm::quantity::fDiffRatio),
                                      0);
+  timeGPUvsCPU_subdet_.initialize(_name,
+                                  "TimeGPUvsCPU",
+                                  hcaldqm::hashfunctions::fSubdet,
+                                  new hcaldqm::quantity::ValueQuantity(hcaldqm::quantity::fCPUtime, true),
+                                  new hcaldqm::quantity::ValueQuantity(hcaldqm::quantity::fGPUtime, true),
+                                  new hcaldqm::quantity::ValueQuantity(hcaldqm::quantity::fN),
+                                  0);
 
   energyGPUvsCPU_subdet_.book(ib, _emap, _subsystem);
   energyDiffGPUCPU_subdet_.book(ib, _emap, _subsystem);
   energyDiffGPUCPU_depth_.book(ib, _emap, _subsystem);
+  timeGPUvsCPU_subdet_.book(ib, _emap, _subsystem);
 }
 
 /* virtual */ void HcalGPUComparisonTask::_resetMonitors(hcaldqm::UpdateFreq uf) { DQTask::_resetMonitors(uf); }
@@ -105,24 +114,36 @@ HcalGPUComparisonTask::HcalGPUComparisonTask(edm::ParameterSet const& ps)
   auto lumiCache = luminosityBlockCache(e.getLuminosityBlock().index());
   _currentLS = lumiCache->currentLS;
 
+  std::map<HcalDetId, double> mRecHitTime;
   std::map<HcalDetId, double> mRecHitEnergy;
 
   for (HBHERecHitCollection::const_iterator it = chbhe_ref->begin(); it != chbhe_ref->end(); ++it) {
     double energy = it->energy();
+    double time = it->time();
 
     //	Explicit check on the DetIds present in the Collection
     HcalDetId did = it->id();
+
+    if (mRecHitTime.find(did) == mRecHitTime.end())
+      mRecHitTime.insert(std::make_pair(did, time));
+    else
+      edm::LogError("HcalGPUComparisonTask") << "Duplicate Rechit from the same HcalDetId";
 
     if (mRecHitEnergy.find(did) == mRecHitEnergy.end())
       mRecHitEnergy.insert(std::make_pair(did, energy));
     else
       edm::LogError("HcalGPUComparisonTask") << "Duplicate Rechit from the same HcalDetId";
-    ;
   }
 
   for (HBHERecHitCollection::const_iterator it = chbhe_target->begin(); it != chbhe_target->end(); ++it) {
     double energy = it->energy();
+    double time = it->time();
     HcalDetId did = it->id();
+
+    if (mRecHitTime.find(did) != mRecHitTime.end()) {
+      timeGPUvsCPU_subdet_.fill(did, mRecHitTime[did], time);
+      mRecHitTime.erase(did);
+    }
 
     if (mRecHitEnergy.find(did) != mRecHitEnergy.end()) {
       energyGPUvsCPU_subdet_.fill(did, mRecHitEnergy[did], energy);

--- a/DataFormats/HcalRecHit/interface/HcalSpecialTimes.h
+++ b/DataFormats/HcalRecHit/interface/HcalSpecialTimes.h
@@ -1,5 +1,6 @@
 #ifndef DataFormats_HcalRecHit_HcalSpecialTimes_h_
 #define DataFormats_HcalRecHit_HcalSpecialTimes_h_
+#include "DataFormats/HcalDetId/interface/HcalSubdetector.h"
 
 // This is an excerpt from QIE10/QIE11 TDC specifications (by T. Zimmerman):
 //
@@ -64,6 +65,73 @@ namespace HcalSpecialTimes {
     constexpr int tdc_code_undershoot = 63;
 
     float t = tdc_to_ns * tdc;
+    if (tdc > six_bits_mask || tdc < 0)
+      t = UNKNOWN_T_INVALID_RANGE;
+    else if (tdc > tdc_code_largestnormal) {
+      // The undershoot code happens by far more often
+      // than any other special code. So check for it first.
+      if (tdc == tdc_code_undershoot)
+        t = UNKNOWN_T_UNDERSHOOT;
+      else if (tdc == tdc_code_overshoot)
+        t = UNKNOWN_T_OVERSHOOT;
+      else if (tdc == tdc_code_invalid)
+        t = UNKNOWN_T_INVALID_CODE;
+      else if (tdc < tdc_code_invalid)
+        t = UNKNOWN_T_50TO57;
+      else
+        t = UNKNOWN_T_DLL_FAILURE;
+    }
+
+    return t;
+  }
+  constexpr inline float getTDCTime(const int tdc, const int subdet, const int depth) {
+    constexpr float tdc_to_ns = 0.5f;
+
+    constexpr int six_bits_mask = 0x3f;
+    constexpr int tdc_code_largestnormal = 49;
+    constexpr int tdc_code_invalid = 58;
+    constexpr int tdc_code_overshoot = 62;
+    constexpr int tdc_code_undershoot = 63;
+
+    float t = -999.f;
+    if (subdet == HcalEndcap) {  // HcalEndcap subdet 2
+      t = tdc_to_ns * tdc;
+    } else if (subdet == HcalBarrel) {  // HcalBarrel subdet 1
+      if (depth == 1) {
+        switch (tdc) {
+          case 0:
+            t = 4.5f;  // prompt
+            break;
+          case 1:
+            t = 11.f;  // slightly delayed
+            break;
+          case 2:
+            t = 18.5f;  // very delayed
+            break;
+          default:
+            t = UNKNOWN_T_INVALID_RANGE;
+            break;
+        }
+      } else {  // Depth 2+
+        switch (tdc) {
+          case 0:
+            t = 3.f;  // prompt
+            break;
+          case 1:
+            t = 8.f;  // slightly delayed
+            break;
+          case 2:
+            t = 17.f;  // very delayed
+            break;
+          default:
+            t = UNKNOWN_T_INVALID_RANGE;
+            break;
+        }
+      }
+    } else {
+      t = UNKNOWN_T_INVALID_RANGE;
+    }
+
     if (tdc > six_bits_mask || tdc < 0)
       t = UNKNOWN_T_INVALID_RANGE;
     else if (tdc > tdc_code_largestnormal) {

--- a/RecoLocalCalo/HcalRecProducers/plugins/alpaka/Mahi.dev.cc
+++ b/RecoLocalCalo/HcalRecProducers/plugins/alpaka/Mahi.dev.cc
@@ -25,14 +25,6 @@ namespace ALPAKA_ACCELERATOR_NAMESPACE {
   namespace hcal::reconstruction {
     namespace mahi {
 
-      ALPAKA_FN_HOST_ACC ALPAKA_FN_INLINE float uint_as_float(uint32_t val) {
-#if defined(__CUDA_ARCH__) or defined(__HIP_DEVICE_COMPILE__)
-        return __uint_as_float(val);
-#else
-        return edm::bit_cast<float>(val);
-#endif
-      }
-
       ALPAKA_FN_HOST_ACC ALPAKA_FN_INLINE uint32_t float_as_uint(float val) {
 #if defined(__CUDA_ARCH__) or defined(__HIP_DEVICE_COMPILE__)
         return __float_as_uint(val);
@@ -535,20 +527,6 @@ namespace ALPAKA_ACCELERATOR_NAMESPACE {
                 auto const dfc = compute_diff_charge_gain(
                     qieType, adc, capid, qieOffsets, qieSlopes, gch < f01HEDigis.size() || gch >= nchannelsf015);
 
-#ifdef COMPUTE_TDC_TIME
-                float tdcTime;
-                if (gch >= f01HEDigis.size() && gch < nchannelsf015) {
-                  tdcTime = HcalSpecialTimes::UNKNOWN_T_NOTDC;
-                } else {
-                  if (gch < f01HEDigis.size())
-                    tdcTime =
-                        HcalSpecialTimes::getTDCTime(tdc_for_sample<Flavor1>(&(f01HEDigis[gch].data()[0]), sample));
-                  else if (gch >= nchannelsf015)
-                    tdcTime = HcalSpecialTimes::getTDCTime(
-                        tdc_for_sample<Flavor3>(&(f3HBDigis.data()[gch - nchannelsf015][0]), sample));
-                }
-#endif  // COMPUTE_TDC_TIME
-
                 // compute method 0 quantities
                 // TODO: need to apply containment
                 // TODO: need to apply time slew
@@ -645,6 +623,7 @@ namespace ALPAKA_ACCELERATOR_NAMESPACE {
               auto const lch = channel.local;
               for (auto i_sample : independent_group_elements_x(acc, nsamplesForCompute)) {
                 auto const sampleWithinWindow = i_sample;
+                auto const sample = i_sample + startingSample;
 
                 int32_t const soi =
                     gch < f01HEDigis.size()
@@ -655,47 +634,35 @@ namespace ALPAKA_ACCELERATOR_NAMESPACE {
                                     ? f01HEDigis[gch].ids()
                                     : (gch < nchannelsf015 ? f5HBDigis.ids()[gch - f01HEDigis.size()]
                                                            : f3HBDigis.ids()[gch - nchannelsf015]);
-                auto const did = HcalDetId{id};
-                auto const hashedId =
-                    did.subdetId() == HcalBarrel
-                        ? did2linearIndexHB(id, mahi.maxDepthHB(), mahi.firstHBRing(), mahi.lastHBRing(), mahi.nEtaHB())
-                        : did2linearIndexHE(id,
-                                            mahi.maxDepthHE(),
-                                            mahi.maxPhiHE(),
-                                            mahi.firstHERing(),
-                                            mahi.lastHERing(),
-                                            mahi.nEtaHE()) +
-                              mahi.offsetForHashes();
 
-                auto const recoParam1 = recoParamsWithPS.recoParamView().param1()[hashedId];
-                auto const recoParam2 = recoParamsWithPS.recoParamView().param2()[hashedId];
-
-                auto const nsamplesToAdd = recoParam1 < 10 ? recoParam2 : (recoParam1 >> 14) & 0xF;
                 // NOTE: must take soi, as values for that thread are used...
                 // NOTE: does not run if soi is bad, because it does not match any sampleWithinWindow
                 if (sampleWithinWindow == static_cast<unsigned>(soi)) {
                   auto const method0_energy = shrMethod0EnergyAccum[lch];
-                  auto const val = shrMethod0EnergySamplePair[lch];
-                  int const max_sample = (val >> 32) & 0xffffffff;
 
-                  float const max_energy = uint_as_float(static_cast<uint32_t>(val & 0xffffffff));
+                  HcalDetId didtmp(id);
+                  auto subdetectorType = didtmp.subdet();
+                  auto subdetectorDepth = didtmp.depth();
 
-                  float const max_energy_1 = static_cast<unsigned>(max_sample) < nsamplesForCompute - 1
-                                                 ? shrEnergyM0PerTS[lch * nsamplesForCompute + max_sample + 1]
-                                                 : 0.f;
-                  float const position = nsamplesToAdd < nsamplesForCompute ? max_sample - soi : max_sample;
-                  auto const sum = max_energy + max_energy_1;
-                  // FIXME: for full comparison with cpu method 0  timing,
-                  // need to correct by slew
-                  // requires an accumulator -> more shared mem -> omit here unless
-                  // really needed
-                  float const time =
-                      max_energy > 0.f && max_energy_1 > 0.f ? 25.f * (position + max_energy_1 / sum) : 25.f * position;
-
+                  float tdcTime = 0.f;
+                  if (gch >= f01HEDigis.size() && gch < nchannelsf015) {
+                    tdcTime = HcalSpecialTimes::UNKNOWN_T_NOTDC;
+                  } else {
+                    if (gch < f01HEDigis.size())
+                      tdcTime =
+                          HcalSpecialTimes::getTDCTime(tdc_for_sample<Flavor1>(f01HEDigis[gch].data().data(), sample),
+                                                       subdetectorType,
+                                                       subdetectorDepth);
+                    else if (gch >= nchannelsf015)
+                      tdcTime = HcalSpecialTimes::getTDCTime(
+                          tdc_for_sample<Flavor3>(f3HBDigis[gch - nchannelsf015].data().data(), sample),
+                          subdetectorType,
+                          subdetectorDepth);
+                  }
                   // store method0 quantities to global mem
                   outputGPU[gch].detId() = id;
                   outputGPU[gch].energyM0() = method0_energy;
-                  outputGPU[gch].timeM0() = time;
+                  outputGPU[gch].timeM0() = tdcTime;
 
                   // check as in cpu version if mahi is not needed
                   // FIXME: KNOWN ISSUE: observed a problem when rawCharge and pedestal
@@ -714,6 +681,38 @@ namespace ALPAKA_ACCELERATOR_NAMESPACE {
 #endif
 
 #ifdef HCAL_MAHI_GPUDEBUG
+                  auto const did = HcalDetId{id};
+                  auto const hashedId =
+                      did.subdetId() == HcalBarrel
+                          ? did2linearIndexHB(
+                                id, mahi.maxDepthHB(), mahi.firstHBRing(), mahi.lastHBRing(), mahi.nEtaHB())
+                          : did2linearIndexHE(id,
+                                              mahi.maxDepthHE(),
+                                              mahi.maxPhiHE(),
+                                              mahi.firstHERing(),
+                                              mahi.lastHERing(),
+                                              mahi.nEtaHE()) +
+                                mahi.offsetForHashes();
+
+                  auto const recoParam1 = recoParamsWithPS.recoParamView().param1()[hashedId];
+                  auto const recoParam2 = recoParamsWithPS.recoParamView().param2()[hashedId];
+                  auto const nsamplesToAdd = recoParam1 < 10 ? recoParam2 : (recoParam1 >> 14) & 0xF;
+                  auto const val = shrMethod0EnergySamplePair[lch];
+                  int const max_sample = (val >> 32) & 0xffffffff;
+
+                  float const max_energy = std::bit_cast<float>(static_cast<uint32_t>(val & 0xffffffff));
+
+                  float const max_energy_1 = static_cast<unsigned>(max_sample) < nsamplesForCompute - 1
+                                                 ? shrEnergyM0PerTS[lch * nsamplesForCompute + max_sample + 1]
+                                                 : 0.f;
+                  float const position = nsamplesToAdd < nsamplesForCompute ? max_sample - soi : max_sample;
+                  auto const sum = max_energy + max_energy_1;
+                  // FIXME: for full comparison with cpu method 0  timing,
+                  // need to correct by slew
+                  // requires an accumulator -> more shared mem -> omit here unless
+                  // really needed
+                  float const time =
+                      max_energy > 0.f && max_energy_1 > 0.f ? 25.f * (position + max_energy_1 / sum) : 25.f * position;
                   printf(" method0_energy = %f max_sample = %d max_energy = %f time = %f\n",
                          method0_energy,
                          max_sample,

--- a/RecoLocalCalo/HcalRecProducers/src/HcalRecHitSoAToLegacy.cc
+++ b/RecoLocalCalo/HcalRecProducers/src/HcalRecHitSoAToLegacy.cc
@@ -56,7 +56,7 @@ void HcalRecHitSoAToLegacy::produce(edm::Event& event, edm::EventSetup const& se
     // build a legacy rechit with the computed detid and MAHI energy
     recHitsLegacy->emplace_back(HcalDetId{rechit.detId()},
                                 rechit.energy(),
-                                0  // timeRising
+                                rechit.timeM0()  // timeRising
     );
     // update the legacy rechit with the Chi2 and M0 values
     recHitsLegacy->back().setChiSquared(rechit.chi2());

